execution/executor/src/components/apply_chunk_output.rs

```diff
@@ -0,0 +1,272 @@
+use scratchpad::SparseMerkleTree;
+use std::{collections::HashMap, convert::TryFrom, sync::Arc};
+use storage_interface::state_view::StateCache;
+
+pub struct ApplyChunkOutput;
+
+impl ApplyChunkOutput {
+    pub fn apply(
+        chunk_output: ChunkOutput,
+        base_accumulator: &Arc<InMemoryAccumulator<TransactionAccumulatorHasher>>,
+    ) -> Result<ExecutedChunk> {
+        let ChunkOutput {
+            state_cache,
+            transactions,
+            transaction_outputs,
+        } = chunk_output;
+
+        // Separate transactions with different VM statuses.
+        let (new_epoch, to_keep, to_discard, to_retry) =
+            Self::sort_transactions(transactions, transaction_outputs)?;
+
+        // Apply the write set, get the latest state.
+        let (account_blobs, roots_with_node_hashes, result_state, next_epoch_state) =
+            Self::apply_write_set(state_cache, new_epoch, &to_keep)?;
+
+        // Calculate TransactionData and TransactionInfo, i.e. the ledger history diff.
+        let (to_commit, transaction_info_hashes) =
+            Self::assemble_ledger_diff(to_keep, account_blobs, roots_with_node_hashes);
+
+        Ok(ExecutedChunk {
+            to_commit,
+            to_discard,
+            to_retry,
+            result_view: ExecutedTrees::new_copy(
+                result_state,
+                Arc::new(base_accumulator.append(&transaction_info_hashes)),
+            ),
+            next_epoch_state,
+            ledger_info: None,
+        })
+    }
+
+    fn sort_transactions(
+        mut transactions: Vec<Transaction>,
+        mut transaction_outputs: Vec<TransactionOutput>,
+    ) -> Result<(
+        bool,
+        Vec<(Transaction, TransactionOutput)>,
+        Vec<Transaction>,
+        Vec<Transaction>,
+    )> {
+        // See if there's a new epoch started
+        let new_epoch_event_key = on_chain_config::new_epoch_event_key();
+        let new_epoch_marker = transaction_outputs
+            .iter()
+            .position(|o| {
+                o.events()
+                    .iter()
+                    .any(|event| *event.key() == new_epoch_event_key)
+            })
+            // Off by one for exclusive index.
+            .map(|idx| idx + 1);
+
+        // Transactions after the epoch ending are all to be retried.
+        let to_retry = if let Some(pos) = new_epoch_marker {
+            transaction_outputs.drain(pos..);
+            transactions.drain(pos..).collect()
+        } else {
+            vec![]
+        };
+
+        // Separate transactions with the Keep status out.
+        let (to_keep, to_discard) =
+            itertools::zip_eq(transactions.into_iter(), transaction_outputs.into_iter())
+                .partition::<Vec<(Transaction, TransactionOutput)>, _>(|(_, o)| {
+                    matches!(o.status(), TransactionStatus::Keep(_))
+                });
+
+        // Sanity check transactions with the Discard status:
+        let to_discard = to_discard
```

Comment:
why do we keep `to_discard`? is it used later?

Reply:
Guess just to be used in `ensure_no_discarded()` for now..